src/gpuarray_buffer_cuda.c

```diff
@@ -484,17 +482,15 @@ static int cuda_write(gpudata *dst, size_t dstoff, const void *src,
 
     cuda_enter(ctx);
 
-    ctx->err = cuEventSynchronize(dst->ev);
-    if (ctx->err != CUDA_SUCCESS) {
-      cuda_exit(ctx);
-      return GA_IMPL_ERROR;
-    }
+    cuda_wait(dst, CUDA_WAIT_WRITE);
 
-    ctx->err = cuMemcpyHtoD(dst->ptr + dstoff, src, sz);
+    ctx->err = cuMemcpyHtoDAsync(dst->ptr + dstoff, src, sz, ctx->s);
```

Comment:
thanks for the detail. You are right, we need to make sure how it work. Otherwise we will need to copy the data to pinned memory ourself.